File: query_primer/query_primer__range_of_score.cs
using static System.Console;
using static System.Math;
using static System.Linq.Enumerable;

internal class Program
{
    static void Main()
    {
        int[] n = ReadLine().Split().Select( int.Parse ).ToArray();
        int sq = (int)Sqrt( n[0] );
        int[] s = new int[n[0]];
        int[] mn = new int[sq];
        int[] mx = new int[sq];

        SetOut( new System.IO.StreamWriter( OpenStandardOutput() ) { AutoFlush = false } );
        foreach( int i in Range( 0, n[0] ) ) {
            s[i] = int.Parse( ReadLine() );
        }
        foreach( int i in Range( 0, sq ) ) {
            foreach( int j in Range( 0, sq ) ) {
                mn[i] = j == 0 ? s[sq * i] : Min( mn[i], s[sq * i + j] );
                mx[i] = j == 0 ? s[sq * i] : Max( mx[i], s[sq * i + j] );
            }
        }
        foreach( int _ in Range( 0, n[1] ) ) {
            int[] lr = ReadLine().Split().Select( x => int.Parse( x ) - 1 ).ToArray();
            int i = lr[0];
            int amax = s[lr[0]];
            int amin = s[lr[0]];

            while( i <= lr[1] ) {
                if( i % sq == 0 && i + sq - 1 <= lr[1] ) {
                    amax = Max( amax, mx[i / sq] );
                    amin = Min( amin, mn[i / sq] );
                    i += sq;
                } else {
                    amax = Max( amax, s[i] );
                    amin = Min( amin, s[i] );
                    i++;
                }
            }
            i = lr[2];

            int bmax = s[lr[2]];
            int bmin = s[lr[2]];

            while( i <= lr[3] ) {
                if( i % sq == 0 && i + sq - 1 <= lr[3] ) {
                    bmax = Max( bmax, mx[i / sq] );
                    bmin = Min( bmin, mn[i / sq] );
                    i += sq;
                } else {
                    bmax = Max( bmax, s[i] );
                    bmin = Min( bmin, s[i] );
                    i++;
                }
            }
            WriteLine( bmax - bmin < amax - amin ? "A" : amax - amin < bmax - bmin ? "B" : "DRAW" );
        }
        Out.Flush();
        return;
    }
}

File: query_primer/query_primer__range_of_score.py
# 実行時間 4510ms
import math
import sys

rl = sys.stdin.readline
n, k = list(map(int, rl().split()))
sq = math.floor(math.sqrt(n))
o = []
s = []
mn = [0 for _ in range(sq)]
mx = [0 for _ in range(sq)]

for _ in [0] * n:
    s.append(int(rl()))
for i in range(sq):
    for j in range(sq):
        mn[i] = s[sq * i] if j == 0 else min(mn[i], s[sq * i + j])
        mx[i] = s[sq * i] if j == 0 else max(mx[i], s[sq * i + j])
for _ in [0] * k:
    al, ar, bl, br = [int(t) - 1 for t in rl().split()]
    i = al
    amn = s[al]
    amx = s[al]

    while i <= ar:
        if i % sq == 0 and i + sq - 1 <= ar:
            if mn[i // sq] < amn:
                amn = mn[i // sq]
            if amx < mx[i // sq]:
                amx = mx[i // sq]
            i += sq
        else:
            if s[i] < amn:
                amn = s[i]
            if amx < s[i]:
                amx = s[i]
            i += 1
    i = bl

    bmn = s[bl]
    bmx = s[bl]

    while i <= br:
        if i % sq == 0 and i + sq - 1 <= br:
            if mn[i // sq] < bmn:
                bmn = mn[i // sq]
            if bmx < mx[i // sq]:
                bmx = mx[i // sq]
            i += sq
        else:
            if s[i] < bmn:
                bmn = s[i]
            if bmx < s[i]:
                bmx = s[i]
            i += 1
    o.append("A" if bmx - bmn < amx - amn else "B" if amx - amn < bmx - bmn else "DRAW")
print("\n".join(o))
